drivers/gpio/gpio_sx1509b.c

```diff
@@ -0,0 +1,372 @@
+ *
+ * @return 0 if successful, failed otherwise
+ */
+static int gpio_sx1509b_config(struct device *dev, int access_op, u32_t pin,
+			       int flags)
+{
+	const struct gpio_sx1509b_config *cfg = dev->config->config_info;
+	struct gpio_sx1509b_drv_data *drv_data = dev->driver_data;
+	struct gpio_sx1509b_pin_state *pins = &drv_data->pin_state;
+	int ret = 0;
+
+	if (flags & GPIO_INT) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	switch (access_op) {
+	case GPIO_ACCESS_BY_PIN:
+		if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
+			pins->dir |= BIT(pin);
+			pins->input_disable &= ~BIT(pin);
+		} else {
+			pins->dir &= ~BIT(pin);
+			pins->input_disable |= BIT(pin);
+		}
+		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+			pins->pull_up |= BIT(pin);
+		} else {
+			pins->pull_up &= ~BIT(pin);
+		}
+		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
+			pins->pull_down |= BIT(pin);
+		} else {
+			pins->pull_down &= ~BIT(pin);
+		}
+		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_NORMAL) {
+			pins->pull_up &= ~BIT(pin);
+			pins->pull_down &= ~BIT(pin);
+		}
+		if (flags & GPIO_DS_DISCONNECT_HIGH) {
+			pins->open_drain |= BIT(pin);
+		} else {
+			pins->open_drain &= ~BIT(pin);
+		}
+		if (flags & GPIO_POL_INV) {
+			pins->polarity |= BIT(pin);
+		} else {
+			pins->polarity &= ~BIT(pin);
+		}
+		break;
+	case GPIO_ACCESS_BY_PORT:
+		if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
+			pins->dir = 0xffff;
+			pins->input_disable = 0x0000;
+		} else {
+			pins->dir = 0x0000;
+			pins->input_disable = 0xffff;
+		}
+		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+			pins->pull_up = 0xffff;
+		} else {
+			pins->pull_up = 0x0000;
+		}
+		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
+			pins->pull_down = 0xffff;
+		} else {
+			pins->pull_down = 0x0000;
+		}
+		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_NORMAL) {
+			pins->pull_up = 0x0000;
+			pins->pull_down = 0x0000;
+		}
+		if (flags & GPIO_DS_DISCONNECT_HIGH) {
+			pins->open_drain = 0xffff;
+		} else {
+			pins->open_drain = 0x0000;
+		}
+		if (flags & GPIO_POL_INV) {
+			pins->polarity = 0xffff;
+		} else {
+			pins->polarity = 0x0000;
+		}
+		break;
+	default:
+		ret = -ENOTSUP;
+		goto out;
+	}
+
+	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
+				    SX1509B_REG_DIR, pins->dir);
+	if (ret)
+		goto out;
+
+	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
+				    SX1509B_REG_INPUT_DISABLE,
+				    pins->input_disable);
+	if (ret)
+		goto out;
+
+	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
+				    SX1509B_REG_PULL_UP,
+				    pins->pull_up);
+	if (ret)
+		goto out;
+
+	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
+				    SX1509B_REG_PULL_DOWN,
+				    pins->pull_up);
+	if (ret)
+		goto out;
+
+	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
+				    SX1509B_REG_OPEN_DRAIN,
+				    pins->pull_up);
```

Comment:
Should the last arg be pins->open_drain?

Reply:
Yes, thanks for pointing that out. Looks like I messed up a merge when working on this earlier.